### PR TITLE
Describe our initial continuous deployment flow

### LINKED ISSUE #96: [4] Determine CI/CD automation process
### Goals
* Determine what we want our automatic deployment rules to be for when we set up CircleCI (#86 )

### Tasks
- [ ] Determine automation rules for where to push and when (esp given multiple devs pushing changes)

### Additional Context
* Related to #86

### PR DESCRIPTION
**Pull request by alexsoble**
## Issue

Closes #96.

## Preview link

👁️ [ars/96/ci-cd-process/docs/how_we_work/deploy_flow.md](https://github.com/HHS/OPRE-OPS/blob/ars/96/ci-cd-process/docs/how_we_work/deploy_flow.md) 👁️ 

4 of the 6 sections are mostly placeholders right now for future questions we will need to answer!

The key section for right now is: [How is code deployed to the development app for testing and initial review?](https://github.com/HHS/OPRE-OPS/blob/ars/96/ci-cd-process/docs/how_we_work/deploy_flow.md#how-is-code-deployed-to-the-development-app-for-testing-and-initial-review)

We can flesh out the other sections as we continue developing the application and our processes.

### CODE REVIEW

**Comment by alexsoble** [comment]
@amymok @carjug Got this started -- let's review and discuss during engineering cowork tomorrow!

**Comment by alexsoble** [comment]
Alrighty! Updated this based on our eng cowork discussion, and added a whole bunch of sections to sketch out the main topics/questions that we'll eventually want covered in this documentation. Will tag the team for review both here and in #96!

**Review by carjug — APPROVED** [review]
Looks good!

File: docs/how_we_work/deploy_flow.md
Comment (carjug) [inline docs/how_we_work/deploy_flow.md:51]: Will POs review changes on dev or staging?

Comment (alexsoble) [reply docs/how_we_work/deploy_flow.md:51]: Good question! I think it depends what level the PO wants to be read-in at. From what I'm seeing so far, I think our PO wants to be read in on a weekly pace rather than a daily pace. So I would say PO reviewing changes on staging would make more sense than on dev. But we can always ask!

File: docs/how_we_work/deploy_flow.md
Comment (amymok) [inline docs/how_we_work/deploy_flow.md:46]: This sentence is really similar to the previous sentence, can we combine or rephrase one of them?

**Review by amymok — CHANGES_REQUESTED** [review]
I am used to the concept of using dev, staging, production to describe the environment, and not to describe the app itself.  The app will need to be tested against the these different environment because there are different configurations as the configurations reflect closer and closer to the production environment.

i.e. This [deployment best practices resource](http://guides.beanstalkapp.com/deployments/best-practices.html)

**Comment by alexsoble** [comment]
Paired with @amymok on editing this!

**Review by amymok — APPROVED** [review]
LGTM!

**Review by ninamak — COMMENTED** [review]
Summarizing some takeaways from our chat about this:

1. Thanks for putting this together for us to react to!
2. Love the flaky and buggy color :)
3. Clarify that we generally don't want to merge a story into Develop in such a buggy state that it might interfere with the review of other work on Develop. 
4. Clarify that before a story is ready for PM review, it has gone through testing on local and that the dev has verified they've met the conditions of acceptance.  The expectation should generally be that relatively few bugs are found.
5. It would be helpful to have a Staging environment where we can be confident that anything on there is ready to be pushed to Production at any time.  i.e., anything on there has gone through all the necessary layers of review.  Similarly, it would be helpful to have a stable environment for our PO and other stakeholders to see or test things on, where they're only seeing things that have gone through our review already and that are waiting their approval.  Therefore, it would be helpful to add a "Demo" environment: work would get reviewed by the PM on Develop, approved by the partner on Demo, and is ready for release on Staging.

**Comment by alexsoble** [comment]
Thank you for the conversation and feedback @ninamak! I updated the doc to better reflect what we discussed. One question we can think about next week — as a PM, would you like to review features in the dev space, or would the demo space be a better fit?

**Comment by ninamak** [comment]
@alexsoble I'd like to review on Dev, and leave Demo as "this has passed all team checks and is waiting on PO/stakeholder feedback."  We can always revisit if things aren't working out the way we expected them to.

**Review by ninamak — APPROVED** [review]
Thanks for making these changes!  LGTM!

**Comment by alexsoble** [comment]
@amymok @carjug Last week Nina and I talked about deploy flow ideas for this PR. Based on Nina's feedback, I updated the proposed flow from {Dev > Staging > Prod} to {Dev > Demo > Staging > Prod}. The differences between Demo and Staging are documented in a lightweight way [here](https://github.com/HHS/OPRE-OPS/blob/ars/96/ci-cd-process/docs/how_we_work/deploy_flow.md#how-is-code-deployed-to-demo-for-further-review). Demo is for stakeholder review; Staging is for approved code that we are ready to ship to Prod at any time.

This PR has three approvals, but I wanted to make sure you had a chance to catch up on the revised flow and ask any questions about Demo/Staging! Let me know what you think.

Feel free to 👍 this comment if the proposed {Dev > Demo > Staging > Prod} flow sounds good to you. And feel free to merge/:ship: if you are the last to review and you approve!

**Comment by carjug** [comment]
@alexsoble the revised deploy flow sounds good to me! 🚢

**Comment by amymok** [comment]
With the added Demo space, I am curious about how this would look like for branching to make it traceable on what is demo site vs what code is pushed there.  Because I am envisioning there may be more than one approved (or is it unapproved?) PR being pushed to the demo space, while staging/production site represent application with all the approved PRs.

**Comment by alexsoble** [comment]
@amymok I think it would make sense to have the code on Demo traceable to a branch, no deploying to Demo manually or via a git tag. The goal of Demo will be to show features to the PO or other stakeholders, so we want the state of Demo to be known and traceable.

**Pull request merged**
